### PR TITLE
Automate artefact collection; build for all VS & clang versions

### PR DESCRIPTION
**Pull request by h-vetinari**
So far, I've just been updating this recipe from the clang side, but I realize now that it's still depending on VS2017, and hasn't yet switched to the new VS2019 default.

But I realised that to be fully usable, this recipe needs to be built for all our available VS versions (as well as all our live clang versions), but that becomes very unwieldy and tiresome with all the hardcoded components/urls.

So I automated this and then expanded the build matrix. I also added myself to maintainers now.

### CODE REVIEW

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -10,12 +21,3 @@ FINAL_CFLAGS:
 
 FINAL_CXXFLAGS:
   - -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -march=k8 -mtune=haswell -ftree-vectorize -fstack-protector-strong -O2 -ffunction-sections -pipe
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml:23]: Not sure if `FINAL_CFLAGS` / `FINAL_CXXFLAGS` is still relevant? I cannot find the corresponding settings in https://github.com/conda-forge/ctng-compilers-feedstock, as the comment above would imply.

**Comment by h-vetinari** [comment]
> (side note - wouldn't we want to build this for VS2019/VS2022 in parallel, like on the vc-feedstock? Or maybe even _in_ the vc-feedstock, if we need to pin so hard to vc- resp. ucrt- versions?)

I'm thinking we should keep this here but perhaps build for the 2x2 matrix of supported (= two latest) VS resp. clang versions?

File: .ci_support/linux_64_.yaml [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
+- x86_64-pc-windows-msvc19.29.30139
 FINAL_CFLAGS:
 - -march=k8 -mtune=haswell -ftree-vectorize -fstack-protector-strong -O2 -ffunction-sections
   -pipe
 FINAL_CXXFLAGS:
 - -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -march=k8 -mtune=haswell
   -ftree-vectorize -fstack-protector-strong -O2 -ffunction-sections -pipe
 MSVC_HEADERS_VERSION:
-- 14.16.27023
+- 14.29.30139
```
Comment (xhochy) [inline .ci_support/linux_64_.yaml:10]: CI is probably failing as the patch version here is different. MS only aligns major and minor with the compiler. Matching compiler and headers will have different patch versions.

Comment (h-vetinari) [reply .ci_support/linux_64_.yaml:10]: Ah, I see there are a bunch more variables being set in `activate-msvc-headers-libs.sh` that now have to be updated as well (and should probably move to the cbc)...

File: recipe/conda_build_config.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/conda_build_config.yaml:23]: This moved to https://github.com/conda-forge/ctng-compiler-activation-feedstock/blob/main/recipe/conda_build_config.yaml; also needs to be updated...

**Comment by h-vetinari** [comment]
@xhochy @isuruf 
After much trial and error, I managed to automate the artefact collection here and generalize it to all VS-versions. The commits are cleaned up to be reviewable individually, though the "meat" of the matter is https://github.com/conda-forge/clang-win-activation-feedstock/pull/18/commits/c83c89a3cdf2706bc3f591568affdfda22578faf.

I've tried to document my understanding of where the various versions come from, and I'd very much appreciate your review on this, as well as the automation of course (though the fact that it works should hopefully speak for itself). Importantly, I'm still inserting hard-coded links into the activation scripts (at installation time), so in principle they run exactly as before.

As far as I understand it, for full functionality we actually need the whole matrix of VS-versions times clang-versions that we have live in conda-forge; I've built this up incrementally: first automate only VS2017, then expand the VS-versions in CBC, then expand the clang matrix as well. You can see that the CI for all these stages is passing -- obviously we have a choice to reduce the matrix as desired.

PTAL

File: recipe/activate-msvc-headers-libs.sh [hunk 3]
```diff
@@ -18,33 +18,19 @@ if [[ ! -d "${MSVC_HEADERS_DIR}" ]]; then
...
+      # which produces a list of "component|url" pairs
+      declare -a COMPONENT_URLS=(
+        @COMPONENT_URLS@
+      )
+
+      for map in "${COMPONENT_URLS[@]}"; do
+        # component=$(echo $map | cut -d '|' -f1)
+        url=$(echo $map | cut -d '|' -f2)
+        # download and unpacks into Contents/VC/Tools/MSVC/${MSVC_HEADERS_VERSION}/...
+        curl -L $url -o tmp.zip && unzip -oqq tmp.zip && rm tmp.zip
```
Comment (h-vetinari) [inline recipe/activate-msvc-headers-libs.sh:31]: After I had scripted all the automation, I ended up not using the component name explicitly in the activation script. I still think it's a value add to keep it, as that makes the activation script way more self-explanatory _with_ the component names in `COMPONENT_URLS`, than if it were just a list of obfuscated URLs.

File: recipe/conda_build_config.yaml [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
+# match vsyear / vsver / cl_version
 # from https://github.com/conda-forge/vc-feedstock/blob/main/recipe/conda_build_config.yaml
 VSYEAR:
   - 2017
 VSVER:
   - 15
-CHOST:
-  - x86_64-pc-windows-msvc19.14.26423
+CL_VERSION:
+  - 19.16.27033
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml]: From what I can tell, these two versions should have been consistent between `CHOST` and the `vs2017_win-64 19.16.27033` run-req in `meta.yaml`; this commit enforces that.

File: recipe/activate-clang_win-64.sh [hunk 5]
```diff
@@ -102,25 +102,26 @@ else
   LIB_USED="${CONDA_PREFIX}/lib"
 fi
 
+CHOST=@CHOST_BASE@@CL_VERSION@
```
Comment (isuruf) [inline recipe/activate-clang_win-64.sh:105]: This line prevents backing up CHOST correctly in _tc_activation. You can just do
```
sed -i.bak "s|@CHOST@|${CHOST_BASE}${CL_VERSION}}|g" ${CHANGE}-${PKG_NAME}.sh
```
in install-pkg.sh

File: recipe/install-pkg.sh [hunk 6]
```diff
@@ -61,8 +60,6 @@ do
     cp ${CHANGE}-${PKG_NAME}.sh ${PREFIX}/etc/conda/${CHANGE}.d/${CHANGE}-${PKG_NAME}.sh
 done
 
-CHOST=$CHOST_BASE$CL_VERSION
-
```
Comment (isuruf) [inline recipe/install-pkg.sh]: Why remove this?

Comment (h-vetinari) [reply recipe/install-pkg.sh]: Oversight while reverting https://github.com/conda-forge/clang-win-activation-feedstock/pull/18/commits/92002c36297480ee069ea34a6072fc8f575afc01, sorry

File: recipe/install-pkg.sh [hunk 7]
```diff
@@ -1,21 +1,67 @@
...
+if [[ "$PKG_NAME" == "msvc-headers-libs" ]]; then
+    ################################################################
+    # create map of required components/urls for msvc-headers-libs #
+    ################################################################
+
+    # download utility to determine necessary download urls
+    curl -LO https://raw.githubusercontent.com/mstorsjo/msvc-wine/master/vsdownload.py
+    # patch it to display the urls in the log
+    patch -u < $RECIPE_DIR/patches/0001-show-payload-url.patch
+    # create temporary folder
```
Comment (isuruf) [inline recipe/install-pkg.sh:13]: Add this as a source in `meta.yaml`?

Comment (h-vetinari) [reply recipe/install-pkg.sh:13]: Fair point, though we'd have to download all of https://github.com/mstorsjo/msvc-wine; I was also aiming to have the script be more easily locally executable for debbuging purposes. In any case, I don't have a strong preference, so I'll give it a shot.

Comment (isuruf) [reply recipe/install-pkg.sh:13]: >  though we'd have to download all of https://github.com/mstorsjo/msvc-wine;

Why? You only download vsdownload.py above

File: recipe/meta.yaml [hunk 8]
```diff
@@ -1,11 +1,21 @@
...
 
 build:
-  number: 0
+  number: 1
+
+source:
+  # not the main source, but a very useful download utility to determine
+  # necessary download urls of windows artefacts, see install-pkg.sh
+  - git_url: https://github.com/mstorsjo/msvc-wine.git
+    git_rev: 635a3aa148a42f701b3e46dd13791a01a3cf87f6
```
Comment (isuruf) [inline recipe/meta.yaml:16]:
```suggestion
  - url: https://raw.githubusercontent.com/mstorsjo/msvc-wine/635a3aa148a42f701b3e46dd13791a01a3cf87f6/vsdownload.py
    sha256: d33edb859d7b17ec1350f2b8c5dca8b7aa39375a4d206e2e8eef3fb95848c0b2
```

File: recipe/install-pkg.sh (hunk 7 above)
Comment (h-vetinari) [reply recipe/install-pkg.sh]: Thanks. I see the linter is unhappy, I'll fix it in few minutes

File: recipe/conda_build_config.yaml [hunk 9]
```diff
@@ -1,21 +1,71 @@
-YEAR:
+# corresponding to the clang versions that are live in
+# https://github.com/conda-forge/clang-compiler-activation-feedstock/blob/main/recipe/conda_build_config.yaml
+clang_version:
+  - 14.0.6
```
Comment (isuruf) [inline recipe/conda_build_config.yaml:4]: Let's not go back to 14

File: recipe/conda_build_config.yaml [hunk 10]
```diff
@@ -1,21 +1,71 @@
...
+  - 15.0.7
+  - 16.0.1
+
+# VS-related things that are zipped together:
+# match vsyear / vsver / cl_version from
+# https://github.com/conda-forge/vc-feedstock/blob/main/recipe/conda_build_config.yaml
+VSYEAR:
+  - 2022
+  - 2019
   - 2017
```
Comment (isuruf) [inline recipe/conda_build_config.yaml:14]: Let's remove 2017

File: .ci_support/linux_64_MSVC_HEADERS_VERSION14.29.30133TOOLCHAIN_COMBINED14.29.16.11clang_version15.0.7.yaml [hunk 11]
```diff
@@ -1,7 +1,5 @@
 CHOST_BASE:
 - x86_64-pc-windows-msvc
-CL_VERSION:
-- 19.29.30139
```
Comment (h-vetinari) [inline .ci_support/linux_64_MSVC_HEADERS_VERSION14.29.30133TOOLCHAIN_COMBINED14.29.16.11clang_version15.0.7.yaml]: It seems that https://github.com/conda-forge/clang-win-activation-feedstock/pull/18/commits/45b2f376dbc9495d832195af8bc4d74e772b253f has affected whether `CL_VERSION` gets picked up for linux/osx. This looks like it should still be there IMO, as its used in `install-pkg.sh`.

File: recipe/conda_build_config.yaml (hunk 9 above)
Comment (h-vetinari) [reply recipe/conda_build_config.yaml]: Fine by me, though it's still the default in conda-forge until https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4215 is merged (hence why I kept it).

File: .ci_support/linux_64_MSVC_HEADERS_VERSION14.29.30133TOOLCHAIN_COMBINED14.29.16.11clang_version15.0.7.yaml (hunk 11 above)
Comment (h-vetinari) [reply .ci_support/linux_64_MSVC_HEADERS_VERSION14.29.30133TOOLCHAIN_COMBINED14.29.16.11clang_version15.0.7.yaml]: Fixed in https://github.com/conda-forge/clang-win-activation-feedstock/pull/18/commits/1673641387e1f24add184ee37d770ba8a34c1fdd

**Comment by isuruf** [comment]
Thanks

**Pull request merged**